resource/aws_ssm_document: Unable to Share with more than 20 Accounts

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform Version

```sh
$ terraform -v
Terraform v0.11.7
+ provider.aws v1.28.0
```

### Affected Resource(s)

* aws_ssm_document

### Terraform Configuration Files

```hcl
locals {
  # 25 randomly generated Account IDs
  account_ids = [
    "118535508916",
    "154276429936",
    "161759918038",
    "194598394645",
    "196052683219",
    "247493144767",
    "250493209165",
    "316287036670",
    "387227595876",
    "425170165879",
    "445725168100",
    "539904583428",
    "590030253021",
    "623181299604",
    "630469292428",
    "636729286444",
    "661009953133",
    "674645499404",
    "689964341709",
    "714964602168",
    "725131050115",
    "751712731274",
    "758789874641",
    "958531182898",
    "977656872983",
  ]
}

resource "aws_ssm_document" "foo" {
  name          = "test_document"
  document_type = "Command"

  permissions {
    type        = "Share"
    account_ids = "${join(",", local.account_ids)}"
  }

  content = <<DOC
  {
    "schemaVersion": "1.2",
    "description": "Check ip configuration of a Linux instance.",
    "parameters": {

    },
    "runtimeConfig": {
      "aws:runShellScript": {
        "properties": [
          {
            "id": "0.aws:runShellScript",
            "runCommand": ["ifconfig"]
          }
        ]
      }
    }
  }
DOC
}
```

### Output

```
Error: Error applying plan:

1 error(s) occurred:

* aws_ssm_document.foo: 1 error(s) occurred:

* aws_ssm_document.foo: [ERROR] Error setting permissions for SSM document: ValidationException: 1 validation error detected: Value '[118535508916, 154276429936, 161759918038, 194598394645, 196052683219, 247493144767, 250493209165, 316287036670, 387227595876, 425170165879, 445725168100, 539904583428, 590030253021, 623181299604, 630469292428, 636729286444, 661009953133, 674645499404, 689964341709, 714964602168, 725131050115, 751712731274, 758789874641, 958531182898, 977656872983]' at 'accountIdsToAdd' failed to satisfy constraint: Member must have length less than or equal to 20
	status code: 400, request id: e3b47fc5-f335-42c6-8e7c-aa1d17780b06
```

### Expected Behavior

* Terraform works within the API limit and is able to share the document with more than 20 Accounts

### Actual Behavior

* Terraform makes a single API call which fails due to the limit

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

* According to AWS Support, they do not currently support raising this limit.

### References

* #5308

--- a/aws/resource_aws_ssm_document.go
+++ b/aws/resource_aws_ssm_document.go
@@ -17,7 +17,8 @@ import (
 )
 
 const (
-	MINIMUM_VERSIONED_SCHEMA = 2.0
+	MINIMUM_VERSIONED_SCHEMA             = 2.0
+	SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT = 20
 )
 
 func resourceAwsSsmDocument() *schema.Resource {
@@ -377,8 +378,6 @@ func setDocumentPermissions(d *schema.ResourceData, meta interface{}) error {
 
 	log.Printf("[INFO] Setting permissions for document: %s", d.Id())
 
-	// Since AccountIdsToRemove has higher priority than AccountIdsToAdd,
-	// we filter out accounts from both lists
 	if d.HasChange("permissions") {
 		o, n := d.GetChange("permissions")
 		oldPermissions := o.(map[string]interface{})
@@ -400,31 +399,25 @@ func setDocumentPermissions(d *schema.ResourceData, meta interface{}) error {
 			}
 		}
 
-		accountIdsToRemove := make([]string, 0)
+		// Since AccountIdsToRemove has higher priority than AccountIdsToAdd,
+		// we filter out accounts from both lists
+		accountIdsToRemove := make([]interface{}, 0)
 		for _, oldPermissionsAccountId := range oldPermissionsAccountIds {
 			if _, contains := sliceContainsString(newPermissionsAccountIds, oldPermissionsAccountId.(string)); !contains {
 				accountIdsToRemove = append(accountIdsToRemove, oldPermissionsAccountId.(string))
 			}
 		}
-		accountIdsToAdd := make([]string, 0)
+		accountIdsToAdd := make([]interface{}, 0)
 		for _, newPermissionsAccountId := range newPermissionsAccountIds {
 			if _, contains := sliceContainsString(oldPermissionsAccountIds, newPermissionsAccountId.(string)); !contains {
 				accountIdsToAdd = append(accountIdsToAdd, newPermissionsAccountId.(string))
 			}
 		}
 
-		input := &ssm.ModifyDocumentPermissionInput{
-			Name:               aws.String(d.Get("name").(string)),
-			PermissionType:     aws.String("Share"),
-			AccountIdsToAdd:    aws.StringSlice(accountIdsToAdd),
-			AccountIdsToRemove: aws.StringSlice(accountIdsToRemove),
-		}
-
-		log.Printf("[DEBUG] Modifying SSM document permissions: %s", input)
-		_, err := ssmconn.ModifyDocumentPermission(input)
-		if err != nil {
+		if err := modifyDocumentPermissions(ssmconn, d.Get("name").(string), accountIdsToAdd, accountIdsToRemove); err != nil {
 			return fmt.Errorf("error modifying SSM document permissions: %s", err)
 		}
+
 	}
 
 	return nil
@@ -478,24 +471,79 @@ func deleteDocumentPermissions(d *schema.ResourceData, meta interface{}) error {
 	log.Printf("[INFO] Removing permissions from document: %s", d.Id())
 
 	permission := d.Get("permissions").(map[string]interface{})
-	var accountsToRemove []*string
+
+	accountIdsToRemove := make([]interface{}, 0)
+
 	if permission["account_ids"] != nil {
-		accountsToRemove = aws.StringSlice([]string{permission["account_ids"].(string)})
-		if strings.Contains(permission["account_ids"].(string), ",") {
-			accountsToRemove = aws.StringSlice(strings.Split(permission["account_ids"].(string), ","))
+
+		if v, ok := permission["account_ids"]; ok && v.(string) != "" {
+			parts := strings.Split(v.(string), ",")
+			accountIdsToRemove = make([]interface{}, len(parts))
+			for i, v := range parts {
+				accountIdsToRemove[i] = v
+			}
+		}
+
+		if err := modifyDocumentPermissions(ssmconn, d.Get("name").(string), nil, accountIdsToRemove); err != nil {
+			return fmt.Errorf("error removing SSM document permissions: %s", err)
+		}
+
+	}
+
+	return nil
+}
+
+func modifyDocumentPermissions(conn *ssm.SSM, name string, accountIdsToAdd []interface{}, accountIdstoRemove []interface{}) error {
+
+	if accountIdsToAdd != nil {
+
+		accountIdsToAddBatch := make([]string, 0, SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT)
+		accountIdsToAddBatches := make([][]string, 0, len(accountIdsToAdd)/SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT+1)
+		for _, accountId := range accountIdsToAdd {
+			if len(accountIdsToAddBatch) == SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT {
+				accountIdsToAddBatches = append(accountIdsToAddBatches, accountIdsToAddBatch)
+				accountIdsToAddBatch = make([]string, 0, SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT)
+			}
+			accountIdsToAddBatch = append(accountIdsToAddBatch, accountId.(string))
+		}
+		accountIdsToAddBatches = append(accountIdsToAddBatches, accountIdsToAddBatch)
+
+		for _, accountIdsToAdd := range accountIdsToAddBatches {
+			_, err := conn.ModifyDocumentPermission(&ssm.ModifyDocumentPermissionInput{
+				Name:            aws.String(name),
+				PermissionType:  aws.String("Share"),
+				AccountIdsToAdd: aws.StringSlice(accountIdsToAdd),
+			})
+			if err != nil {
+				return err
+			}
 		}
 	}
 
-	permInput := &ssm.ModifyDocumentPermissionInput{
-		Name:               aws.String(d.Get("name").(string)),
-		PermissionType:     aws.String("Share"),
-		AccountIdsToRemove: accountsToRemove,
-	}
+	if accountIdstoRemove != nil {
 
-	_, err := ssmconn.ModifyDocumentPermission(permInput)
+		accountIdsToRemoveBatch := make([]string, 0, SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT)
+		accountIdsToRemoveBatches := make([][]string, 0, len(accountIdstoRemove)/SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT+1)
+		for _, accountId := range accountIdstoRemove {
+			if len(accountIdsToRemoveBatch) == SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT {
+				accountIdsToRemoveBatches = append(accountIdsToRemoveBatches, accountIdsToRemoveBatch)
+				accountIdsToRemoveBatch = make([]string, 0, SSM_DOCUMENT_PERMISSIONS_BATCH_LIMIT)
+			}
+			accountIdsToRemoveBatch = append(accountIdsToRemoveBatch, accountId.(string))
+		}
+		accountIdsToRemoveBatches = append(accountIdsToRemoveBatches, accountIdsToRemoveBatch)
 
-	if err != nil {
-		return fmt.Errorf("Error removing permissions for SSM document: %s", err)
+		for _, accountIdsToRemove := range accountIdsToRemoveBatches {
+			_, err := conn.ModifyDocumentPermission(&ssm.ModifyDocumentPermissionInput{
+				Name:               aws.String(name),
+				PermissionType:     aws.String("Share"),
+				AccountIdsToRemove: aws.StringSlice(accountIdsToRemove),
+			})
+			if err != nil {
+				return err
+			}
+		}
+
 	}
 
 	return nil

--- a/aws/resource_aws_ssm_document_test.go
+++ b/aws/resource_aws_ssm_document_test.go
@@ -108,6 +108,26 @@ func TestAccAWSSSMDocument_permission_private(t *testing.T) {
 	})
 }
 
+func TestAccAWSSSMDocument_permission_batching(t *testing.T) {
+	name := acctest.RandString(10)
+	ids := "123456789012,123456789013,123456789014,123456789015,123456789016,123456789017,123456789018,123456789019,123456789020,123456789021,123456789022,123456789023,123456789024,123456789025,123456789026,123456789027,123456789028,123456789029,123456789030,123456789031,123456789032"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSSSMDocumentDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSSSMDocumentPrivatePermissionConfig(name, ids),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSSSMDocumentExists("aws_ssm_document.foo"),
+					resource.TestCheckResourceAttr(
+						"aws_ssm_document.foo", "permissions.type", "Share"),
+				),
+			},
+		},
+	})
+}
+
 func TestAccAWSSSMDocument_permission_change(t *testing.T) {
 	name := acctest.RandString(10)
 	idsInitial := "123456789012,123456789013"

